http/native_server.ts

```
@@ -0,0 +1,579 @@
+  }
+
+  /**
+   * Send a response to the request.
+   *
+   * @param {Response|Promise<Response>} response Response to the request.
+   * @throws {Deno.errors.BadResource} When the response has already been sent.
+   */
+  respondWith(response: Response | Promise<Response>): Promise<void> {
+    if (this.#done) {
+      throw ERROR_RESPONSE_SENT;
+    }
+
+    this.#resolver(response);
+    this.#done = true;
+
+    return this.#responsePromise;
+  }
+}
+
+/**
+ * Options for running a server.
+ */
+export interface ServerInit {
+  /**
+   * Optionally specifies the TCP address to listen on, in the form
+   * `host:port`. If not provided, ":80" is used by default for HTTP and ":443"
+   * is used by default for HTTPS.
+   */
+  addr?: string;
+
+  /**
+   * The handler for individual HTTP requests.
+   */
+  handler: Handler;
+}
+
+export class Server {
+  addr?: string;
+  handler: Handler;
+  #closed = false;
+  #listeners: Set<Deno.Listener> = new Set();
+  #httpConnections: Set<Deno.HttpConn> = new Set();
+
+  /**
+   * Constructs a new Server instance.
+   *
+   * @param {ServerInit} options Options for running a server.
+   */
+  constructor({ addr, handler }: ServerInit) {
+    this.addr = addr;
+    this.handler = handler;
+  }
+
+  /**
+   * Accept incoming connections on the given listener, and handle requests on
+   * these connections with the given handler.
+   *
+   * Throws a server closed error if called after the server has been closed.
+   *
+   * Will always close the listener.
+   *
+   * @param {Deno.Listener} listener The listener to accept connections from.
+   * @throws {Deno.errors.Http} When the server has been closed.
+   */
+  async serve(listener: Deno.Listener): Promise<void> {
+    if (this.#closed) {
+      throw ERROR_SERVER_CLOSED;
+    }
+
+    this.trackListener(listener);
+
+    const mux = this.serveMux(listener);
+
+    try {
+      for await (const requestEvent of mux) {
+        requestEvent.respondWith(this.handler(
```

Should we swallow errors from here as issues with a particular request, close that connection, but otherwise not kill the process? Seems to be what go's http lib does?

---

Have gone for this - 2aac74b.